SetupDataPkg/ConfApp/SetupConf.c

```diff
@@ -168,6 +168,122 @@ PrintOptions (
+  EFI_STATUS           Status          = EFI_SUCCESS;
+  CHAR16               *AlignedVarName = NULL;
+  CHAR16               *UnalignedVarName;
+  EFI_GUID             *Guid;
+  UINT32               Attributes;
+  CHAR8                *Data;
+  UINT32               CRC32;
+  UINT32               LenToCRC32;
+  UINT32               CalcCRC32 = 0;
+  CONFIG_VAR_LIST_HDR  *VarList;
+  UINT32               ListIndex = 0;
+
+  if ((Value == NULL) || (ValueSize == 0)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  while (ListIndex < ValueSize) {
+    // index into variable list
+    VarList = (CONFIG_VAR_LIST_HDR *)(Value + ListIndex);
+
+    if (ListIndex + sizeof (*VarList) + VarList->NameSize + VarList->DataSize + sizeof (*Guid) +
+        sizeof (Attributes) + sizeof (CRC32) > ValueSize)
+    {
+      // the NameSize and DataSize have bad values and are pushing us past the end of the binary
+      DEBUG ((DEBUG_ERROR, "SVD settings had bad NameSize or DataSize, unable to process all settings\n"));
+      break;
+    }
+
+    /*
+     * Var List is in DmpStore format:
+     *
+     *  struct {
+     *    RUNTIME_VAR_LIST_HDR VarList;
+     *    CHAR16 Name[VarList->NameSize/2];
+     *    EFI_GUID Guid;
+     *    UINT32 Attributes;
+     *    CHAR8 Data[VarList->DataSize];
+     *    UINT32 CRC32; // CRC32 of all bytes from VarList to end of Data
+     *  }
+     */
+    UnalignedVarName = (CHAR16 *)(VarList + 1);
+    Guid             = (EFI_GUID *)((CHAR8 *)UnalignedVarName + VarList->NameSize);
+    Attributes       = *(UINT32 *)(Guid + 1);
+    Data             = (CHAR8 *)Guid + sizeof (*Guid) + sizeof (Attributes);
+    CRC32            = *(UINT32 *)(Data + VarList->DataSize);
+    LenToCRC32       = sizeof (*VarList) + VarList->NameSize + VarList->DataSize + sizeof (*Guid) + sizeof (Attributes);
+
+    // on next iteration, skip past this variable
+    ListIndex += LenToCRC32 + sizeof (CRC32);
+
+    // validate CRC32
+    CalcCRC32 = CalculateCrc32 (VarList, LenToCRC32);
+
+    if (CalcCRC32 != CRC32) {
+      // CRC didn't match, drop this variable, continue to the next
+      DEBUG ((DEBUG_ERROR, "SVD setting failed CRC check, skipping applying setting: %a Status: %r \
+      Received CRC32: %u Calculated CRC32: %u\n", UnalignedVarName, Status, CRC32, CalcCRC32));
+      continue;
+    }
+
+    AlignedVarName = AllocatePool (VarList->NameSize);
+    if (AlignedVarName == NULL) {
+      Status = EFI_OUT_OF_RESOURCES;
+      return Status;
+    }
+
+    // Copy non-16 bit aligned UnalignedVarName to AlignedVarName, not using StrCpyS functions as they assert on non-16 bit alignment
+    CopyMem (AlignedVarName, UnalignedVarName, VarList->NameSize);
+
+    // first delete the variable in case of size or attributes change, not validated here as this is only allowed
+    // in manufacturing mode. Don't retrieve the status, if we fail to delete, try to write it anyway. If we fail
+    // there, just log it and move on
+    gRT->SetVariable (

```

Comment:
Probably want to check if the return is good.

Reply:
Here I figured we don't care. Either the delete doesn't work, in which case we can still try the variable set and it may fail and we'll record it. Either way, we'll keep processing down the variable list.

Reply:
Okay, this is fine.